src/System.Web.OData/OData/Routing/DefaultODataPathHandler.cs

```diff
@@ -112,12 +112,10 @@ private ODataPath Parse(string serviceRoot, string odataPath, IServiceProvider r
             {
                 Contract.Assert(serviceRoot != null);
 
-                serviceRootUri = new Uri(
-                    serviceRoot.EndsWith("/", StringComparison.Ordinal)
-                        ? serviceRoot
-                        : serviceRoot + "/");
+                string serviceRootStr = serviceRoot.EndsWith("/", StringComparison.Ordinal) ? serviceRoot : serviceRoot + "/";
+                serviceRootUri = new Uri(serviceRootStr);
 
-                fullUri = new Uri(serviceRootUri, odataPath);
+                fullUri = new Uri(serviceRootStr + odataPath);
```

Comment:
Can you add a test for long Uris?

Reply:
This change will require adding a test.

Reply:
Sorry, I'm a rookie. I don't know how to write this test.

Reply:
Take a look at the tests in DefaultODataPathHandlerTest.cs. Many of them call parse just like you would need to, you just need to construct a long uri similar to the issue you have been seeing.